hipe: change mfa_info_table lock to rwmutex

mikpe opened the pull request:
The hipe_mfa_info_table lock is currently a mutex, causing
unnecessary contention and serialization of lookups for
apply-like constructs.

- change the lock from a mutex to an rwmutex
- split hipe_get_na_nofail_locked into a "try" path which
  only needs a read-lock, and a "slow" path which requires
  a write-lock
- reimplement hipe_get_na_nofail (used by apply ops etc) to
  first call the "try" path with a read lock, and if that
  fails the "slow" path with a write lock
- reimplement hipe_get_na_nofail_locked (used by maintenance
  code) to call the "try" path and then optionally the "slow"
  path, but without taking locks since its caller already has
  the write lock
- adjust remaining lock ops to take/release full write-lock
- change hipe_mfa_get_trampoline to call get_locked not put_locked,
  allowing it to take a read lock instead of a write lock

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

sverker commented:
Nice.

I think hipe_find_mfa_from_ra() also will suffice with a read lock. It looks like a pure read operation.

If you are editing, I think it would be nice for the functions that demand write lock to be renamed
from *_locked to *_rwlocked.

And feel free to squash into one commit and force push your branch to github.

mikpe commented:
Changed hipe_find_mfa_from_ra() to only take a read lock, changed suffix of functions requiring write lock to _rwlocked, squashed the three commits, force-pushed.

OTP-Maintainer commented:
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------
